File: python/day00.py
"""Day 0: The Tyranny of the Rocket Equation (tutorial dry run).

Python algorithm reference mirroring src/day00.ml, kept as a second
implementation to cross-check the OCaml answers. Input is one integer module
mass per line.
"""

import sys


def parse_input(raw: str) -> list[int]:
    return [int(line) for line in raw.splitlines() if line.strip()]


def fuel(mass: int) -> int:
    """Base equation: floor(mass / 3) - 2."""
    return mass // 3 - 2


def total_fuel(mass: int) -> int:
    """Part 2: fuel for the fuel, iterated until a step is <= 0."""
    total = 0
    f = fuel(mass)
    while f > 0:
        total += f
        f = fuel(f)
    return total


def part1(parsed: list[int]) -> str:
    return str(sum(fuel(mass) for mass in parsed))


def part2(parsed: list[int]) -> str:
    return str(sum(total_fuel(mass) for mass in parsed))


def solve(raw: str) -> tuple[str, str]:
    parsed = parse_input(raw)
    return part1(parsed), part2(parsed)


def main(argv: list[str]) -> int:
    path = argv[1] if len(argv) > 1 else "inputs/day00.txt"
    try:
        with open(path, encoding="utf-8") as handle:
            raw = handle.read()
    except OSError as error:
        print(f"Could not read {path}: {error}", file=sys.stderr)
        print(
            f"Create {path} locally (it is gitignored) or pass an explicit path.",
            file=sys.stderr,
        )
        return 1
    one, two = solve(raw)
    print("day00")
    print(f"  part1: {one}")
    print(f"  part2: {two}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: src/day00.ml
(* Day 0: The Tyranny of the Rocket Equation (tutorial dry run).

   Input is one integer module mass per line. This day is the pipeline smoke
   test: it exercises parse -> part1 -> part2 -> registry -> bench -> tests once,
   end to end, against a puzzle whose answers are already locked in
   ../aoc2020_Prolog (src/day00.pl). *)

type parsed = string list
(* [Day_registry] fixes [parsed] as [string list] for every day, so
   [parse_input] keeps the raw lines and [masses] does the int conversion. *)

let parse_input : string -> parsed = Io.nonempty_lines

(* [String.trim] is what makes this robust to Windows CRLF inputs; without it
   [int_of_string] would choke on a trailing carriage return. *)
let masses lines = List.map (fun line -> int_of_string (String.trim line)) lines

(* Part 1 kernel, straight from the spec: floor (mass / 3) - 2. Masses are
   positive, so OCaml's truncating [/] agrees with flooring here. *)
let fuel mass = (mass / 3) - 2

(* Part 2: fuel has mass, so it needs its own fuel, and so on. A fixed-point
   iteration -- apply [fuel] to its own output, accumulate each positive step,
   and stop once a step asks for nothing positive. *)
let rec total_fuel mass =
  let step = fuel mass in
  if step <= 0 then 0 else step + total_fuel step

let sum_by f lines = masses lines |> List.fold_left (fun acc mass -> acc + f mass) 0
let part1 lines = string_of_int (sum_by fuel lines)
let part2 lines = string_of_int (sum_by total_fuel lines)

(* Parse once, produce both answers -- the shape every later day reuses. *)
let solve raw =
  let parsed = parse_input raw in
  (part1 parsed, part2 parsed)
